TDataFrame.hxx

```diff
@@ -22,828 +22,1042 @@
+};
+using FilterBasePtr_t = std::shared_ptr<TDataFrameFilterBase>;
+using FilterBaseVec_t = std::vector<FilterBasePtr_t>;
 
-template<typename FilterF, typename PrevDataFrame>
-class TDataFrameFilter
-   : public TDataFrameFilterBase,
-     public TDataFrameInterface<TDataFrameFilter<FilterF, PrevDataFrame>>
-{
-   template<typename A, typename B> friend class TDataFrameAction;
-   template<typename A, typename B> friend class TDataFrameFilter;
-   template<typename A, typename B> friend class TDataFrameBranch;
-   template<typename A> friend class TDataFrameInterface;
-   using BranchTypes = typename FunctionTraits<FilterF>::ArgTypes;
-   using TypeInd = typename gens<BranchTypes::size>::type;
-   using TDFInterface = TDataFrameInterface<TDataFrameFilter<FilterF, PrevDataFrame>>;
+template <typename FilterF, typename PrevDataFrame>
+class TDataFrameFilter : public TDataFrameFilterBase {
+   using BranchTypes_t = typename Internal::TDFTraitsUtils::TFunctionTraits<FilterF>::ArgTypes_t;
+   using TypeInd_t = typename Internal::TDFTraitsUtils::TGenS<BranchTypes_t::fgSize>::Type_t;
 
 public:
-   TDataFrameFilter(FilterF f, const BranchList& bl, PrevDataFrame& pd)
-      : fFilter(f), fBranches(bl), fTmpBranches(pd.fTmpBranches), fPrevData(pd),
-        fFirstData(pd.fFirstData) { }
+   TDataFrameFilter(FilterF f, const BranchList &bl, std::weak_ptr<PrevDataFrame> pd)
+      : fFilter(f), fBranches(bl), fTmpBranches(pd.lock()->fTmpBranches), fPrevData(pd.lock().get()),
+        fFirstData(pd.lock()->fFirstData) { }
 
-   TDataFrame& GetDataFrame() const {
-      return fFirstData;
-   }
+   std::weak_ptr<TDataFrameImpl> GetDataFrame() const { return fFirstData; }
 
-   TDataFrameFilter(const TDataFrameFilter&) = delete;
+   TDataFrameFilter(const TDataFrameFilter &) = delete;
 
-private:
-   bool CheckFilters(unsigned slot, int entry) {
-      if(entry != fLastCheckedEntry[slot]) {
-         if(!fPrevData.CheckFilters(slot, entry)) {
+   bool CheckFilters(unsigned int slot, int entry)
+   {
+      if (entry != fLastCheckedEntry[slot]) {
+         if (!fPrevData->CheckFilters(slot, entry)) {
             // a filter upstream returned false, cache the result
             fLastResult[slot] = false;
          } else {
             // evaluate this filter, cache the result
-            fLastResult[slot] = CheckFilterHelper(BranchTypes(), TypeInd(), slot, entry);
+            fLastResult[slot] = CheckFilterHelper(BranchTypes_t(), TypeInd_t(), slot, entry);
          }
          fLastCheckedEntry[slot] = entry;
       }
       return fLastResult[slot];
    }
 
-   template<int... S, typename... BranchTypes>
-   bool CheckFilterHelper(TypeList<BranchTypes...>, seq<S...>, unsigned slot, int entry) {
+   template <int... S, typename... BranchTypes>
+   bool CheckFilterHelper(Internal::TDFTraitsUtils::TTypeList<BranchTypes...>,
+                          Internal::TDFTraitsUtils::TSeq<S...>,
+                          unsigned int slot, in
```

Comment:
we don't *really* need to `enable_shared_from_this`: `TDataFrame{Branch,Filter,Impl}` could have a method `GetFirstData` that returns `fFirstData` for the first two classes and `this` for `TDataFrameImpl`.